tests/test-umockdev-vala.vala

```diff
@@ -855,6 +855,114 @@ t_mt_uevent ()
   assert_cmpuint (change_count, CompareOperator.EQ, num_changes);
 }
 
+static bool
+ioctl_custom_handle_ioctl_cb(UMockdev.IoctlBase handler, UMockdev.IoctlClient client)
+{
+    if (client.request == 1) {
+        client.complete(*(long*)client.arg.data, 0);
+    } else if (client.request == 2) {
+        client.complete(-1, Posix.ENOMEM);
+    } else if (client.request ==3 ) {
+        var data = client.arg.resolve(0, sizeof(int));
+
+        *(int*) data.data = (int) 0xc00fffee;
+
+        client.complete(0, 0);
+    } else {
+        client.complete(-5, 1);
+    }
+    return true;
+}
+
+static bool
+ioctl_custom_handle_write_cb(UMockdev.IoctlBase handler, UMockdev.IoctlClient client)
+{
+    /* NOTE:
+     *
+     * This code uses g_object_{get,set}_data for simplicty. Any realy code
+     * should likely subclass IoctlBase, override the vfunc's and keep its
+     * state in there (this only works if you do not need a per-client state).
+     */
+    client.set_data("written", client.arg);
+    client.complete(client.arg.data.length, 0);
+    return true;
+}
+
+static bool
+ioctl_custom_handle_read_cb(UMockdev.IoctlBase handler, UMockdev.IoctlClient client)
+{
+    UMockdev.IoctlData written = client.steal_data("written");
+
+    /* Return EAGAIN if nothing has been written. */
+    if (written == null) {
+        client.complete(-1, Posix.EAGAIN);
+        return true;
+    }
+
+    /* Return the first n bytes of the buffer, then discard the rest. Real code
+     * would likely need to track the offset.
+     */
+    ssize_t read_len = ssize_t.min(written.data.length, client.arg.data.length);
+
+    Posix.memcpy(client.arg.data, written.data, read_len);
+
+    client.complete(read_len, 0);
+
+    return true;
+}
+
+void
+t_ioctl_custom()
+{
+  var tb = new UMockdev.Testbed ();
+
+  tb_add_from_string (tb, """P: /devices/test
+N: test
+E: SUBSYSTEM=test
+""");
+
+  var handler = new UMockdev.IoctlBase();
+  var write_buf = new uint8[10] { 1, 2, 3, 4 ,5 ,6 ,7, 8, 9, 10 };
```

Comment:
The comma/spaces are still wrong here. (That was already part of the previous round, but already marked resolved)